The beginning of hunk CLI IDs

jonathantanmy2 opened the pull request:
Right now you can use them in `rub` to assign uncommitted hunks, but that's it. No other operations, no displaying, and so on - those are left for future work. But sending this PR to make sure I'm on the right track.

jonathantanmy2 committed:
id: deduplicate uncommitted files

The input to `IdMap`'s constructor may contain multiple
`HunkAssignments` for one uncommitted file, so teach it to deduplicate
them. Committed files do not have the same problem, as can be seen from
the included test (but the test tests committed files anyway to prevent
regressions).

The test uses `status`, which is not ideal (this is an `IdMap` concern),
but I have included a TODO describing how and when it can be removed.

jonathantanmy2 committed:
id, rub: uncommitted hunk to stack

This is the beginning of hunk CLI IDs. As of this commit, the user
can use a CLI ID to assign an uncommitted hunk to a stack if they know
what the hunk's ID is (right now, there is no way other than knowing
GitButler internals; a "diff" or "show" command that presents hunk CLI
IDs to the user is left for future work).

Note that `uncommitted_and_committed_file_cli_ids` in
crates/but/tests/but/command/status.rs should not be deleted yet, since
we don't have tests for committed hunk CLI IDs yet.

jonathantanmy2 committed:
id: parse IDs with more than two characters

An `if` block assumes that IDs are only two characters, but this is no
longer true. Fix that.

jonathantanmy2 committed:
rub: combine all `makes_no_sense_error` cases

A subsequent commit will introduce a new `CliId` enum variant. To avoid
even more clutter in `handle`, combine all `makes_no_sense_error` cases
into one.

Byron commented on crates/but/tests/but/command/rub.rs:
```diff
@@ -21,3 +21,63 @@ Rubbed the wrong way. Source 'nonexistent1' not found. If you just performed a G
+        .assert()
+        .success()
+        .stdout_eq(str![[r#"
+Assigned a hunk in [1ma.txt[0m → [32m[A][0m.
```
Might it be 'nicer' to use the `file![]` syntax here to get a colored SVG?

Byron reviewed (commented):
Thanks a lot! I love how easy it seems to add hunk-assignment with the current infrastructure.

I just had a few comments, but wonder if unit or integration tests for Id make it feasible to test some property of hunks more directly? Right now these are only tested on `but` CLI integration level.

Thanks again.

jonathantanmy2 commented on crates/but/tests/but/command/status.rs:
```diff
@@ -187,3 +187,81 @@ fn json_shows_paths_as_strings() -> anyhow::Result<()> {
+// TODO This test demonstrates how IDs are assigned to uncommitted and committed
+// files that have multiple hunks. This test can be removed when we have CLI
+// IDs for hunks, a command (e.g. `rub`) is taught to use them, and that command
+// is tested.
```
Already commented on this in the commit message.

jonathantanmy2 commented on crates/but/src/id/mod.rs:
```diff
@@ -561,6 +585,15 @@ pub enum CliId {
+    UncommittedHunk {
+        /// The hunk header.
+        hunk_header: Option<HunkHeader>,
+        /// The path.
```
> I don't think we need to do comments if they don't add to what's obvious to the field.

I agree, but Clippy requires comments due to `#![forbid(missing_docs)]`. But following your comment below, I've added a comment linking to `HunkAssignment` which explains what `None` means.

I've renamed `path_bytes` to `path`.

Byron commented on crates/but/src/id/tests.rs:
```diff
@@ -225,6 +244,40 @@ fn non_commit_ids_do_not_collide() -> anyhow::Result<()> {
     ]
     "#);
+
+    // uncommitted hunks do not collide with anything else
```
Something I am still struggling with is the lack of a traversal over all available ShortIDs, so one has to somehow know the Id to request.
I might look into a test-tool like that.

Byron committed:
refactor

- add tests I found important
- add test utilities
- changed ordering in `resolve_entity_to_ids()` as it didn't seem important
  due to lack of test or usecase in code.

Byron reviewed (commented):
Thanks a lot, I am loving where this is going!

I will merge after working through the unit-tests to make them answer a couple of questions I have been having.

Byron commented:
I am out of time for now but would still want to give the integration tests a good look. My hunch here is that we are in copy-paste mode with a lot of changes to the fixture after creating it, and I wonder if this should ever be done for the IdMap which should be fully unit-testable.
Probably it's nothing though, just the typical struggle of trying to find just the right tests for the 'high' level they are at.

Byron committed:
Address Copilot review

Byron commented on crates/but/src/id/mod.rs:
```diff
@@ -337,6 +341,22 @@ impl IdMap {
+                .cmp(&b.stack_id)
+                .then_with(|| a.path_bytes.cmp(&b.path_bytes))
+                .then_with(|| a.hunk_header.cmp(&b.hunk_header))
+        });
```
- It doesn't feel that way, and my problem here is that I don't know why this sorting is necessary. Removing it instead until there is a test for it.
CC @jonathantanmy2 in case that's a bad idea.

Byron commented on crates/but/src/id/tests.rs:
```diff
@@ -380,5 +513,98 @@ mod util {
+                        .expect("BUG: valid ID means no error")
+                })
+                .sorted()
+                .collect()
```
I think it's clear enough as these are only locally sorted.

Byron commented on crates/but/src/command/legacy/rub/assign.rs:
```diff
@@ -120,28 +146,36 @@ pub(crate) fn branch_name_to_stack_id(
+    Ok(assignments
+        .into_iter()
+        .filter(move |assignment| assignment.path_bytes == path)
+        .map(move |assignment| (assignment.hunk_header, assignment.path_bytes)))
```
I don't get it, but removed an unnecessary clone.

Byron commented on crates/but/src/id/mod.rs:
```diff
@@ -569,6 +601,7 @@ pub enum CliId {
         id: ShortId,
     },
     /// A commit in the workspace identified by its SHA.
+    // TODO: Ensure our prefixes are unique within the set of known commits.
```
Made the comment clearer and added a test to show current, problematic behaviour.

Byron commented:
I completely misunderstood the integration tests which are indeed testing commands, so there isn't anything to adjust right now. My general feeling is that it would be nice if scenarios could be more tailor-made to start out as, but maybe overall it's just most convenient to start at one of a few initial states and alter it to be the actual start state.
